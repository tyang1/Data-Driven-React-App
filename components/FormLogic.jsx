import React from "react";
import { isFunction, isObject } from "../utils/typeCheck.js";

const validationLogic = {
  email: email => {
    return `email validation! ${email}`;
  },
  password: pwd => {
    let text = pwd.toString();
    const isValid = /^(?=.*[\d])(?=.*[A-Z])(?=.*[a-z])(?=.*[!@#$%^&*])[\w!@#$%^&*]{8,}$/.test(
      text
    );
    if (isValid) {
      return true;
    } else {
      return false;
    }
  }
};

function createLogic() {
  return Object.assign(
    {},
    {
      logicStore: {},
      set(logic) {
        this.logicStore = Object.assign({}, this.logicStore, logic);
      },
      get(logic) {
        return this.logicStore[logic];
      }
    }
  );
}

export const FormLogicContext = React.createContext(createLogic());

export class FormLogic extends React.Component {
  constructor(props) {
    super(props);
  }
  render() {
    const { logic } = this.props;
    return (
      <FormLogicContext.Provider value={logic}>
        {this.props.children}
      </FormLogicContext.Provider>
    );
  }
}

//TODO: to hook up to the actual backend services
export async function login(name, password) {
  // const { token } = await fetchCreate("authentication", { name, password });
  // rememberToken(token);
}

//
export function withLogic(mapLogicToProps) {
  //Map logic to props for children component, here we need the validation function as props to the wrappedComponent;
  let mappedCompProps = {};
  if (!isObject(mapLogicToProps) && !isFunction(mapLogicToProps)) {
    throw new Error("withLogic expects an argument and it must be a function");
  }

  for (let [key, value] of Object.entries(mapLogicToProps)) {
    mappedCompProps[value] = validationLogic[key];
  }
  return WrappedComponent => {
    class withLogic extends React.Component {
      render() {
        return <WrappedComponent {...mappedCompProps} />;
      }
    }
    return withLogic;
  };
}
