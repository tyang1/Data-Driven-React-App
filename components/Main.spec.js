import React from "react";
import { FormLogicContext } from "./FormLogic.jsx";
import { graphql, QueryRenderer } from "react-relay";
import ReactTestRenderer from "react-test-renderer";
// import { configure } from "enzyme";
// import Adapter from "enzyme-adapter-react-16";
// import { mount, shallow } from "enzyme";
import { createMockEnvironment, MockPayloadGenerator } from "relay-test-utils";

// configure({ adapter: new Adapter() });

const getLogicWithFormLogicContext = (
  context = { logicStore: [], set: () => null, get: () => null }
) => {
  // Will then mock the FormLogic module being used in our Main component
  jest.mock("./FormLogic.jsx", () => {
    return {
      FormLogicContext: {
        Consumer: props => props.children(null)
      }
    };
  });
  return require("./Main").Main;
};

describe("<Main/>", () => {
  it("should render component", () => {
    const Main = getLogicWithFormLogicContext();
    const environment = createMockEnvironment();
    const TestRenderer = () => (
      <QueryRenderer
        environment={environment}
        query={graphql`
          query MainSpecQuery @relay_test_operation {
            store {
              ...Main_store
            }
          }
        `}
        variables={{}}
        render={({ error, props }) => {
          if (props) {
            return <Main store={props.store} />;
          } else if (error) {
            return error.message;
          }
          return "Loading...";
        }}
      />
    );
    const renderer = ReactTestRenderer.create(<TestRenderer />);
    environment.mock.resolveMostRecentOperation(operation =>
      MockPayloadGenerator.generate(operation)
    );

    expect(renderer).toMatchSnapshot();
  });
});
