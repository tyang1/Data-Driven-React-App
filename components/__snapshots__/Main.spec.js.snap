// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main/> should render component 1`] = `
<div
  className="MainContainer"
>
  <h3>
    Links
  </h3>
  <form
    onSubmit={[Function]}
  >
    <input
      placeholder="Title"
      type="text"
    />
    <input
      placeholder="Url"
      type="text"
    />
    <button
      type="submit"
    >
      Add
    </button>
  </form>
  <input
    onChange={[Function]}
    placeholder="Search"
    type="text"
  />
  <select
    onChange={[Function]}
  >
    <option
      value="1"
    >
      1
    </option>
    <option
      selected={true}
      value="10"
    >
      10
    </option>
  </select>
  <ul
    className="linkLayout"
  />
</div>
`;
